src/main/scala/org/graphframes/GraphFrame.scala

```diff
@@ -32,76 +32,188 @@ import org.graphframes.pattern._
+ * @groupname structure Structure information
+ * @groupname conversions Conversions
+ * @groupname stdlib Standard graph algorithms
+ * @groupname degree Graph topology
+ * @groupname motif Motif finding
  */
-class GraphFrame protected(
-    @transient val vertices: DataFrame,
-    @transient val edges: DataFrame) extends Logging with Serializable {
+class GraphFrame private(
+    @transient private val _vertices: DataFrame,
+    @transient private val _edges: DataFrame) extends Logging with Serializable {
 
   import GraphFrame._
 
-  require(vertices.columns.contains(ID),
-    s"Vertex ID column '$ID' missing from vertex DataFrame, which has columns: "
-      + vertices.columns.mkString(","))
-  require(edges.columns.contains(SRC),
-    s"Source vertex ID column '$SRC' missing from edge DataFrame, which has columns: "
-      + edges.columns.mkString(","))
-  require(edges.columns.contains(DST),
-    s"Destination vertex ID column '$DST' missing from edge DataFrame, which has columns: "
-      + edges.columns.mkString(","))
-
-  private[graphframes] def sqlContext: SQLContext = vertices.sqlContext
-
   /** Default constructor is provided to support serialization */
   protected def this() = this(null, null)
 
+  // ============== Basic structural methods ============
+
+  /**
+   * The dataframe representation of the vertices of the graph.
+   *
+   * It contains a column called [[GraphFrame.ID]] with the id of the vertex,
+   * and various other user-defined attributes with other attributes.
+   *
+   * The order of the columns is available in [[vertexColumns]].
+   *
+   * @group structure
+   */
+  def vertices: DataFrame = {
+    if (_vertices == null) {
+      throw new Exception("You cannot use GraphFrame objects within a Spark closure")
+    }
+    _vertices
+  }
+
+  /**
+   * The dataframe representation of the edges of the graph.
+   *
+   * It contains two columns called [[GraphFrame.SRC]] and [[GraphFrame.DST]] that contain
+   * the ids of the source vertex and the destination vertex of each edge, respectively.
+   * It may also contain various other columns with user-defined attributes for each edge.
+   *
+   * For symmetric graphs, both pairs src -> dst and dst -> src are present with the same
+   * attributes for each pair.
+   *
+   * The order of the columns is available in [[edgeColumns]].
+   *
+   * @group structure
+   */
+  // TODO(tjh) mention what happens with multiple edges.
+  def edges: DataFrame = {
+    if (_edges == null) {
+      throw new Exception("You cannot use GraphFrame objects within a Spark closure")
+    }
+    _edges
+  }
+
+  /**
+   * The column names in the [[vertices]] DataFrame, in order.
+   *
+   * Helper method for [[toGraphX]] which specifies the schema of vertex attributes.
+   * The vertex attributes of the returned [[Graph.vertices]] are given as a [[Row]],
+   * and this method defines the column ordering in that [[Row]].
+   *
+   * @group structure
```

Comment:
These 4 items should be in group conversions since they are only useful alongside toGraphX